diff --git a/include/zephyr/net/net_pkt.h b/include/zephyr/net/net_pkt.h
index e1091056852..6ba05bb688e 100644
--- a/include/zephyr/net/net_pkt.h
+++ b/include/zephyr/net/net_pkt.h
@@ -1903,6 +1903,10 @@ const char *net_pkt_pool2str(struct net_buf_pool *pool);
 #define net_pkt_print(...)
 #endif /* CONFIG_NET_DEBUG_NET_PKT_ALLOC */
 
+#if CONFIG_NET_PKT_LOG_LEVEL >= LOG_LEVEL_DBG
+void net_pkt_debug_set_trace_slab(struct k_mem_slab *slab);
+#endif /* CONFIG_NET_PKT_LOG_LEVEL >= LOG_LEVEL_DBG */
+
 /* New allocator, and API are defined below.
  * This will be simpler when time will come to get rid of former API above.
  */
diff --git a/subsys/net/ip/net_pkt.c b/subsys/net/ip/net_pkt.c
index 79036bd5d3e..47319c4a78b 100644
--- a/subsys/net/ip/net_pkt.c
+++ b/subsys/net/ip/net_pkt.c
@@ -518,6 +518,15 @@ static inline struct net_buf_pool *get_data_pool(struct net_context *context)
 #define get_data_pool(...) NULL
 #endif /* CONFIG_NET_CONTEXT_NET_PKT_POOL */
 
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+static struct k_mem_slab *net_pkt_trace_slab;
+
+void net_pkt_debug_set_trace_slab(struct k_mem_slab *slab)
+{
+	net_pkt_trace_slab = slab;
+}
+#endif /* NET_LOG_LEVEL >= LOG_LEVEL_DBG */
+
 #if NET_LOG_LEVEL >= LOG_LEVEL_DBG
 void net_pkt_unref_debug(struct net_pkt *pkt, const char *caller, int line)
 {
@@ -559,6 +568,12 @@ void net_pkt_unref(struct net_pkt *pkt)
 
 #if NET_LOG_LEVEL >= LOG_LEVEL_DBG
 #if CONFIG_NET_PKT_LOG_LEVEL >= LOG_LEVEL_DBG
+	if (net_pkt_trace_slab && pkt->slab == net_pkt_trace_slab && ref == 1) {
+		NET_DBG("TRACE_SLAB [%d] pkt %p ref %ld frags %p (%s():%d)",
+			k_mem_slab_num_free_get(pkt->slab),
+			pkt, ref - 1, pkt->frags, caller, line);
+	}
+
 	NET_DBG("%s [%d] pkt %p ref %ld frags %p (%s():%d)",
 		slab2str(pkt->slab), k_mem_slab_num_free_get(pkt->slab),
 		pkt, ref - 1, pkt->frags, caller, line);
@@ -640,6 +655,12 @@ struct net_pkt *net_pkt_ref(struct net_pkt *pkt)
 	} while (!atomic_cas(&pkt->atomic_ref, ref, ref + 1));
 
 #if CONFIG_NET_PKT_LOG_LEVEL >= LOG_LEVEL_DBG
+	if (net_pkt_trace_slab && pkt && pkt->slab == net_pkt_trace_slab) {
+		NET_DBG("TRACE_SLAB_REF [%d] pkt %p ref %ld (%s():%d)",
+			k_mem_slab_num_free_get(pkt->slab),
+			pkt, ref + 1, caller, line);
+	}
+
 	NET_DBG("%s [%d] pkt %p ref %ld (%s():%d)",
 		slab2str(pkt->slab), k_mem_slab_num_free_get(pkt->slab),
 		pkt, ref + 1, caller, line);
@@ -1528,6 +1549,10 @@ static struct net_pkt *pkt_alloc(struct k_mem_slab *slab, k_timeout_t timeout)
 
 #if NET_LOG_LEVEL >= LOG_LEVEL_DBG
 	net_pkt_alloc_add(pkt, true, caller, line);
+	if (net_pkt_trace_slab && slab == net_pkt_trace_slab) {
+		NET_DBG("TRACE_SLAB_ALLOC [%d] pkt %p (%s():%d)",
+			k_mem_slab_num_free_get(slab), pkt, caller, line);
+	}
 #endif
 
 	net_pkt_cursor_init(pkt);
